File: fc-chat/demo/part3/ch4/message-system-add-push/src/main/java/net/prostars/messagesystem/handler/websocket/FetchChannelInviteCodeRequestHandler.java
package net.prostars.messagesystem.handler.websocket;

import net.prostars.messagesystem.constant.IdKey;
import net.prostars.messagesystem.constant.MessageType;
import net.prostars.messagesystem.dto.domain.UserId;
import net.prostars.messagesystem.dto.websocket.inbound.FetchChannelInviteCodeRequest;
import net.prostars.messagesystem.dto.websocket.outbound.ErrorResponse;
import net.prostars.messagesystem.dto.websocket.outbound.FetchChannelInviteCodeResponse;
import net.prostars.messagesystem.service.ChannelService;
import net.prostars.messagesystem.service.ClientNotificationService;
import org.springframework.stereotype.Component;
import org.springframework.web.socket.WebSocketSession;

@Component
@SuppressWarnings("unused")
public class FetchChannelInviteCodeRequestHandler
    implements BaseRequestHandler<FetchChannelInviteCodeRequest> {

  private final ChannelService channelService;
  private final ClientNotificationService clientNotificationService;

  public FetchChannelInviteCodeRequestHandler(
      ChannelService channelService, ClientNotificationService clientNotificationService) {
    this.channelService = channelService;
    this.clientNotificationService = clientNotificationService;
  }

  @Override
  public void handleRequest(WebSocketSession senderSession, FetchChannelInviteCodeRequest request) {
    UserId senderUserId = (UserId) senderSession.getAttributes().get(IdKey.USER_ID.getValue());

    if (!channelService.isJoined(request.getChannelId(), senderUserId)) {
      clientNotificationService.sendMessage(
          senderSession,
          senderUserId,
          new ErrorResponse(
              MessageType.FETCH_CHANNEL_INVITECODE_REQUEST, "Not joined the channel."));
      return;
    }

    channelService
        .getInviteCode(request.getChannelId())
        .ifPresentOrElse(
            inviteCode ->
                clientNotificationService.sendMessage(
                    senderSession,
                    senderUserId,
                    new FetchChannelInviteCodeResponse(request.getChannelId(), inviteCode)),
            () ->
                clientNotificationService.sendMessage(
                    senderSession,
                    senderUserId,
                    new ErrorResponse(
                        MessageType.FETCH_CHANNEL_INVITECODE_REQUEST,
                        "Fetch channel invite code failed.")));
  }
}

File: fc-chat/demo/part3/with-push/src/main/kotlin/com/yscorp/withpush/messagesystem/handler/websocket/FetchChannelInviteCodeRequestHandler.kt
package com.yscorp.withpush.messagesystem.handler.websocket

import com.yscorp.withpush.messagesystem.constant.IdKey
import com.yscorp.withpush.messagesystem.constant.MessageType
import com.yscorp.withpush.messagesystem.dto.domain.UserId
import com.yscorp.withpush.messagesystem.dto.websocket.inbound.FetchChannelInviteCodeRequest
import com.yscorp.withpush.messagesystem.dto.websocket.outbound.ErrorResponse
import com.yscorp.withpush.messagesystem.service.ChannelService
import com.yscorp.withpush.messagesystem.service.ClientNotificationService
import org.springframework.stereotype.Component
import org.springframework.web.socket.WebSocketSession

@Component
@Suppress("unused")
class FetchChannelInviteCodeRequestHandler(
    private val channelService: ChannelService,
    private val clientNotificationService: ClientNotificationService
) :
    BaseRequestHandler<FetchChannelInviteCodeRequest> {

    override fun handleRequest(senderSession: WebSocketSession, request: FetchChannelInviteCodeRequest) {
        val senderUserId: UserId = senderSession.attributes[IdKey.USER_ID.value] as UserId

        if (!channelService.isJoined(request.channelId, senderUserId)) {
            clientNotificationService.sendMessage(
                senderSession,
                senderUserId,
                ErrorResponse(
                    MessageType.FETCH_CHANNEL_INVITECODE_REQUEST, "Not joined the channel."
                )
            )
            return
        }

        val inviteCode1 = channelService
            .getInviteCode(request.channelId)

        if (inviteCode1 == null) {
            clientNotificationService.sendMessage(
                senderSession,
                senderUserId,
                ErrorResponse(
                    MessageType.FETCH_CHANNEL_INVITECODE_REQUEST, "Channel invite code not found."
                )
            )
            return
        } else {
            clientNotificationService.sendMessage(
                senderSession,
                senderUserId,
                ErrorResponse(
                    MessageType.FETCH_CHANNEL_INVITECODE_REQUEST,
                    "Fetch channel invite code failed."
                )
            )
        }

    }
}
